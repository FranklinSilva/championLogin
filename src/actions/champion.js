import api from '../services/api';
import deviceStorage from '../services/deviceStorage';

export const getChampionInfo = (idToken) =>{

    return async (dispatch) => {
        return api({
            method: 'get',
            url: "user",
            headers: {
                "Accept": "application/vnd.github.v3+json",
                'Authorization': "token " + idToken
                //"client_id": '2bf4709c19409f01e27d'
            },
            })  
            .then((response) => {
                console.log(response);
                dispatch(getChampionInfoSuccess(response.data, idToken));
            })
            .catch((error) => {
                // Handle returned errors here
                console.log(error)
                //apiUtils.handleError(error);
                console.log(error.response);
                console.log(error);
                dispatch();
            });
    }
}

export const getChampionInfoSuccess = (champion, idToken) => {
    if(!!idToken)
        champion.idToken = idToken;
    
    deviceStorage.saveItem("champion", JSON.stringify(champion));
    
    return {
        type: 'GET_CHAMPION_INFO',
        payload:{
            champion: champion
        }
    }
}
