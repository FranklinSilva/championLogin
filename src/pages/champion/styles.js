import {Platform, StyleSheet} from 'react-native';
import Variables from '../../styles'

const styles = StyleSheet.create({
    container: {
        backgroundColor: Variables.colors.lighter,
        flex:1
    },
    header: { 
        backgroundColor: Variables.colors.primary,
        borderBottomWidth: 0.3,
        borderColor: Variables.colors.gray,
        justifyContent: "center",
        alignItems: "center",
        paddingVertical: 10,
        paddingHorizontal: 25,
        marginTop: 0,
        flexDirection: 'row'
    },
    headerText: {
        textAlign: "center",
        color: Variables.colors.white,
        fontSize: 18,
        fontWeight: "500"
    },
    infoBoxLine: {
        flexDirection: "row",
        marginTop: 10
    },
    imageBox:{
        marginTop: 30,
        alignSelf: "center",
    },
    boxRepositories: {
        padding: 10,
        backgroundColor: Variables.colors.primaryDarken,
        alignItems: 'center',
        justifyContent: 'center',
        height: 60,
        width: 100,
        borderRadius: 6,
        marginRight: 15
    },
    boxGists: {
        padding: 10,
        backgroundColor: Variables.colors.primaryLighter,
        alignItems: 'center',
        justifyContent: 'center',
        height: 60,
        width: 100,
        borderRadius: 6,
        marginLeft: 15
    },
    boxText: {
        textAlign: "center",
        color: Variables.colors.white,
        fontSize: 14,
        fontWeight: "500"
    },
    infoText: {
        textAlign: "center",
        color: Variables.colors.dark,
        fontSize: 14,
        fontWeight: "500"
    },
    infoTextName: {
        textAlign: "center",
        color: Variables.colors.black,
        fontSize: 16,
        fontWeight: "800",
        marginTop: 10
    },
});

export default styles;