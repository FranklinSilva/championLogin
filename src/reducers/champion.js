export default function champion(state={}, action) {

    switch (action.type) {
        case 'GET_CHAMPION_INFO':
            console.log(action);
            return action.payload.champion;
        default:
            return state;
        
    }
}