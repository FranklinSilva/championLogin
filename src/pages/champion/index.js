import React, { Component } from 'react';
import {View, Text, TouchableOpacity, ScrollView, Image} from 'react-native';
import styles from './styles';
import Variables from '../../styles';
import { revoke } from 'react-native-app-auth';

//redux things
import {connect} from 'react-redux';

import * as championActions from '../../actions/champion';
import { bindActionCreators } from 'redux';
import deviceStorage from '../../services/deviceStorage';

//Unsafe
import {CLIENT_ID, CLIENT_sECRET} from '../../services/constants';

class Champion extends Component {
    static navigationOptions = {
        header: null
      }

    constructor(props){
        super(props);
        this.state = {
            list: [],
            isLoading: false,
            fetchFailed: false,
            champion: {}
        }
    }

 

    componentDidMount(){
        console.log('this.props')
        this.fetchUser(this.props.champion);
    }

    fetchUser = (champion) => {
        this.setState({isLoading: false, champion})
    }

    logout = (url) => {
        deviceStorage.clearStorage();
        this.props.navigation.goBack();
        if(!!url){
            this.props.navigation.state.params.revokeToken(url);
        }
    }

    revokeToken = async () => {
        const config = {
            redirectUrl: 'com.characterlist://oauthredirect',
            clientId: CLIENT_ID,
            clientSecret: CLIENT_sECRET,
            scopes: ['identity'],
            serviceConfiguration: {
              authorizationEndpoint: 'https://github.com/login/oauth/authorize',
              tokenEndpoint: 'https://github.com/login/oauth/access_token',
              revocationEndpoint:
                'https://github.com/settings/connections/applications/' + CLIENT_ID
            }
          };


        const result = await revoke(config, {
            tokenToRevoke: this.state.champion.idToken,
            includeBasicAuth: true,
            sendClientId: true,
        });

        console.log(result)

        this.logout(result.url);
    }

    render() {
        return (
        <View style={styles.container}>
            <View style={styles.header}>
                <View style={{flex:1}}></View>
                <Text style={styles.headerText}>Your Information</Text>
                <TouchableOpacity style={{flex:1, alignItems: 'flex-end'}} onPress={() => this.logout()}>
                    <Image
                        style={{ 
                        height: 17,
                        width: 17, borderRadius: 8}}
                        resizeMode="contain"
                        source={require("../../images/exit.png")}/>
                </TouchableOpacity>
            </View>
          
            <ScrollView style={{backgroundColor: Variables.colors.lighter}} contentContainerStyle={{justifyContent: "center",
                alignItems: "center"}}>
                    <View style={styles.imageBox}>
                        <Image
                        style={{ 
                        height: 150,
                        width: 200, borderRadius: 8}}
                        resizeMode="contain"
                        source={{uri:this.state.champion.avatar_url}}/>

                        <Text style={styles.infoTextName}>{this.state.champion.name}</Text>
                        <Text style={styles.infoText}>{this.state.champion.company}</Text>
                        <Text style={styles.infoText}>{this.state.champion.email}</Text>
                    </View>

                    <View style={styles.infoBoxLine}>
                        <View style={styles.boxRepositories}>
                            <Text style={styles.boxText}>{this.state.champion.public_repos}</Text>
                            <Text style={styles.boxText}>Repositories</Text>
                        </View>
                        <View style={styles.boxGists}>
                            <Text style={styles.boxText}>{this.state.champion.public_gists}</Text>
                            <Text style={styles.boxText}>Gists</Text>
                        </View>
                    </View>
                    
                    <TouchableOpacity style={styles.revokeButton} onPress={() => this.revokeToken()}>
                        <Text style={styles.revokeText}>Logout and Revoke Token</Text>
                    </TouchableOpacity>

            </ScrollView>
        </View>
        );
    }

}

const mapStateToProps = state => ({
    champion: state.champion,
})

const mapDispatchToProps = dispatch => 
bindActionCreators({...championActions}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Champion);
