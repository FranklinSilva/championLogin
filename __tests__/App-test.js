/**
 * @format
 */

import 'react-native';
import React from 'react';
import Welcome from '../src/pages/welcome';
import Champion from '../src/pages/champion';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import reducer from '../src/reducers';
import {getChampionInfo, getChampionInfoSuccess} from '../src/actions/champion';

// Note: test renderer must be required after react-native.
import renderer, { create, act } from 'react-test-renderer';

//Mocks
jest.mock('@react-navigation/core')

jest.mock('@react-native-community/async-storage', () => {
  return {
    getItem: async (...args) => args,
    setItem: async (...args) => args,
    removeItem: async (...args) => args,
  };
});

jest.mock('react-navigation', () => ({
  withNavigation: (Component) => (props) => (
    <Component navigation={{ navigate: jest.fn() }} {...props} />
  ),
}))

const store = createStore(reducer, applyMiddleware(thunk))


/* WELCOME PAGE */
const welcome = create(
<Provider store={store}>
  <Welcome />
</Provider>);

it('welcome renders correctly', () => {
  expect(welcome).toMatchSnapshot();
});

it('Login Github Press', () => {
  //press the button
  const githubButton = welcome.root.findByProps({testID: 'gitHubLogin'}).props;
  act(() => githubButton.onPress());
})


it('Get User Info', () => {
  expect(getChampionInfo('idToken123')).toBeDefined();
  expect(getChampionInfoSuccess({},'idToken123')).toBeDefined();
})

/* CHAMPION PAGE */
const champion = create(
  <Provider store={store}>
    <Champion />
  </Provider>);

it('champion renders correctly', () => {
  expect(champion).toMatchSnapshot();
});
