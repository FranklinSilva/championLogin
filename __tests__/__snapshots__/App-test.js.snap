// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`champion renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#eee",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#032f62",
        "borderBottomWidth": 0.3,
        "borderColor": "#a9a9a9",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 0,
        "paddingHorizontal": 25,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 18,
          "fontWeight": "500",
          "textAlign": "center",
        }
      }
    >
      Your Information
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
          "opacity": 1,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/images/exit.png",
          }
        }
        style={
          Object {
            "borderRadius": 8,
            "height": 17,
            "width": 17,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    style={
      Object {
        "backgroundColor": "#eee",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "marginTop": 30,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "borderRadius": 8,
              "height": 150,
              "width": 200,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#454F63",
              "fontSize": 16,
              "fontWeight": "800",
              "marginTop": 10,
              "textAlign": "center",
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#666",
              "fontSize": 14,
              "fontWeight": "500",
              "textAlign": "center",
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#666",
              "fontSize": 14,
              "fontWeight": "500",
              "textAlign": "center",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": 10,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#05264c",
              "borderRadius": 6,
              "height": 60,
              "justifyContent": "center",
              "marginRight": 15,
              "padding": 10,
              "width": 100,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 14,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          />
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 14,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          >
            Repositories
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#065ec5",
              "borderRadius": 6,
              "height": 60,
              "justifyContent": "center",
              "marginLeft": 15,
              "padding": 10,
              "width": 100,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 14,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          />
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 14,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          >
            Gists
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#e37a7a",
            "borderRadius": 6,
            "marginTop": 30,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 14,
              "fontWeight": "500",
              "textAlign": "center",
            }
          }
        >
          Logout and Revoke Token
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`welcome renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#eee",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#032f62",
        "borderBottomWidth": 0.3,
        "borderColor": "#a9a9a9",
        "justifyContent": "center",
        "marginTop": 0,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 18,
          "fontWeight": "500",
          "textAlign": "center",
        }
      }
    >
      Champion Login
    </Text>
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 14,
          "fontWeight": "500",
          "textAlign": "center",
        }
      }
    >
      Access your Github information
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginTop": 0,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/images/Octocat.png",
        }
      }
      style={
        Object {
          "height": 150,
          "width": 200,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#666",
            "fontSize": 14,
            "marginLeft": 5,
          }
        }
      >
        Auto login has failed
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#010b18",
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
            "width": 180,
          }
        }
        testID="gitHubLogin"
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 12,
              "textAlign": "center",
            }
          }
        >
          Login with your Github Account
        </Text>
      </View>
    </View>
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
      transparent={true}
      visible={false}
    >
      <View
        accessible={true}
        focusable={true}
        forwardedRef={[Function]}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "black",
            "bottom": 0,
            "height": 750,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 750,
          }
        }
      />
      <View
        forwardedRef={[Function]}
        hideModalContentWhileAnimating={false}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        pointerEvents="box-none"
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
            "margin": 0,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
          }
        }
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "margin": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": undefined,
                "flex": 1,
                "height": 750,
                "justifyContent": "space-between",
                "paddingVertical": 0,
                "width": 750,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#032f62",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingHorizontal": 20,
                  "paddingVertical": 10,
                  "width": 750,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 16,
                  }
                }
              >
                Revoke you access
              </Text>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#fff",
                      "fontSize": 14,
                      "textAlign": "center",
                    }
                  }
                >
                  Close
                </Text>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flex": 1,
                    "overflow": "hidden",
                  },
                  undefined,
                ]
              }
            >
              <RNCWebView
                cacheEnabled={true}
                injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
                injectedJavaScriptForMainFrameOnly={true}
                javaScriptEnabled={true}
                messagingEnabled={false}
                onContentProcessDidTerminate={[Function]}
                onHttpError={[Function]}
                onLoadingError={[Function]}
                onLoadingFinish={[Function]}
                onLoadingProgress={[Function]}
                onLoadingStart={[Function]}
                onMessage={[Function]}
                onShouldStartLoadWithRequest={[Function]}
                source={
                  Object {
                    "uri": undefined,
                  }
                }
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Object {
                      "backgroundColor": "#ffffff",
                    },
                    Object {
                      "flex": 10,
                      "marginTop": 0,
                    },
                  ]
                }
                useSharedProcessPool={true}
              />
            </View>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
